File: packages/typescript/src/options.ts
// SPDX-License-Identifier: Apache-2.0

import { XenoteerError } from "./errors.js";
import type { SdkErrorCode } from "./errors.js";
import type {
  ReconnectPolicy,
  WebSocketFactory,
} from "./events.js";
import type { ProtocolRange } from "./protocol.generated.js";

export type TokenProvider = () => string | Promise<string>;
export type TokenSource = string | TokenProvider;

export type SafeLogOperation =
  | "http.request"
  | "artifact.upload"
  | "artifact.download"
  | "artifact.delete"
  | "websocket.handshake";

export type SafeLogOutcome = "started" | "succeeded" | "failed";

export type SafeLogRoute =
  | "/v1/status"
  | "/v1/ws"
  | "/v1/artifacts"
  | "/v1/artifacts/:artifact_id"
  | "/v1/desktops/:desktop_id/commands"
  | "/v1/desktops/:desktop_id/commands/:command_id"
  | "/v1/desktops/:desktop_id/lease"
  | "/v1/desktops/:desktop_id/lease/:lease_id/renew"
  | "/v1/desktops/:desktop_id/windows"
  | "/v1/desktops/:desktop_id/windows/query"
  | "/v1/desktops/:desktop_id/windows/resolve"
  | "/v1/desktops/:desktop_id/windows/:window_reference"
  | "/v1/desktops/:desktop_id/windows/wait"
  | "/v1/desktops/:desktop_id/accessibility/elements/list"
  | "/v1/desktops/:desktop_id/accessibility/elements/query"
  | "/v1/desktops/:desktop_id/accessibility/elements/resolve"
  | "/v1/desktops/:desktop_id/accessibility/elements/snapshot"
  | "/v1/desktops/:desktop_id/accessibility/elements/wait"
  | "/v1/desktops/:desktop_id/clipboard/read"
  | "/v1/desktops/:desktop_id/screenshots"
  | "/v1/desktops/:desktop_id/viewer-tickets"
  | "unknown";

export interface SafeLogEvent {
  readonly operation: SafeLogOperation;
  readonly outcome: SafeLogOutcome;
  readonly attempt?: number;
  readonly method?: "GET" | "POST" | "DELETE";
  readonly route?: SafeLogRoute;
  readonly status?: number;
  readonly requestBytes?: number;
  readonly responseBytes?: number;
  readonly errorCode?: SdkErrorCode;
}

export type SafeLogHook = (event: Readonly<SafeLogEvent>) => void;

const TOKEN68 = /^[A-Za-z0-9._~+/-]+={0,}$/u;
const inspectSymbol = Symbol.for("nodejs.util.inspect.custom");

/** An opaque bearer credential whose string, JSON, and inspect forms are redacted. */
export class BearerToken {
  readonly #value: string;

  constructor(value: string) {
    if (
      value.length < 32
      || value.length > 1024
      || !TOKEN68.test(value)
      || /=[^=]/u.test(value)
    ) {
      throw new XenoteerError("invalid_token", "invalid Xenoteer bearer token");
    }
    this.#value = value;
  }

  authorizationHeader(): string {
    return `Bearer ${this.#value}`;
  }

  toString(): string {
    return "BearerToken(<redacted>)";
  }

  toJSON(): string {
    return "<redacted>";
  }

  [inspectSymbol](): string {
    return this.toString();
  }
}

export interface XenoteerClientOptions {
  readonly baseUrl: string;
  readonly token: TokenSource;
  readonly requestTimeoutMs?: number;
  readonly maxResponseBytes?: number;
  readonly maxArtifactBytes?: number;
  readonly fetch?: typeof fetch;
  /**
   * Receives metadata-only diagnostics. Request/response bodies, credentials,
   * lease IDs, artifact IDs, viewer tickets, and server-provided prose are
   * deliberately never included.
   */
  readonly log?: SafeLogHook;
  readonly clientName?: string;
  readonly clientVersion?: string;
  readonly protocolRange?: ProtocolRange;
  /** Header-capable WebSocket adapter retained for initial and reconnect attempts. */
  readonly webSocketFactory?: WebSocketFactory;
  readonly reconnect?: ReconnectPolicy;
  readonly webSocketQueueCapacity?: number;
  readonly webSocketMaxMessageBytes?: number;
  readonly webSocketHandshakeTimeoutMs?: number;
  readonly webSocketAcknowledgmentTimeoutMs?: number;
  readonly webSocketHeartbeatGraceMs?: number;
  readonly webSocketCloseTimeoutMs?: number;
}

function validateInteger(
  value: number,
  minimum: number,
  maximum: number,
  label: string,
): number {
  if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
    throw new XenoteerError("invalid_request", `${label} is outside its supported range`);
  }
  return value;
}

function validateDescriptor(value: unknown, label: string): string {
  if (
    typeof value !== "string"
    || new TextEncoder().encode(value).byteLength < 1
    || new TextEncoder().encode(value).byteLength > 128
    || /[\u0000-\u001f\u007f-\u009f]/u.test(value)
  ) {
    throw new XenoteerError(
      "invalid_request",
      `${label} must be 1..128 UTF-8 bytes without control characters`,
    );
  }
  return value;
}

function snapshotProtocolRange(
  value: ProtocolRange | undefined,
): Readonly<ProtocolRange> | undefined {
  if (value === undefined) return undefined;
  if (
    typeof value !== "object"
    || value === null
    || !Number.isInteger(value.major)
    || !Number.isInteger(value.minMinor)
    || !Number.isInteger(value.maxMinor)
    || value.major !== 1
    || value.minMinor < 0
    || value.minMinor > 65_535
    || value.maxMinor < 0
    || value.maxMinor > 65_535
    || value.minMinor > value.maxMinor
  ) {
    throw new XenoteerError(
      "invalid_request",
      "protocolRange must be a valid version-one minor range",
    );
  }
  return Object.freeze({
    major: value.major,
    minMinor: value.minMinor,
    maxMinor: value.maxMinor,
  });
}

function snapshotReconnect(
  reconnect: ReconnectPolicy | undefined,
): Readonly<ReconnectPolicy> | undefined {
  if (reconnect === undefined) return undefined;
  const maxAttempts = validateInteger(
    reconnect.maxAttempts ?? 5,
    0,
    100,
    "reconnect maxAttempts",
  );
  const initialDelayMs = validateInteger(
    reconnect.initialDelayMs ?? 100,
    1,
    60_000,
    "reconnect initial delay",
  );
  const maxDelayMs = validateInteger(
    reconnect.maxDelayMs ?? 5_000,
    1,
    120_000,
    "reconnect maximum delay",
  );
  if (initialDelayMs > maxDelayMs) {
    throw new XenoteerError("invalid_request", "reconnect delays are reversed");
  }
  return Object.freeze({ maxAttempts, initialDelayMs, maxDelayMs });
}

/**
 * Takes the one retained connection-policy snapshot before any network I/O.
 * Callback references are borrowed for the client lifetime; mutable numeric
 * policy and protocol objects are copied and frozen.
 *
 * @internal
 */
export function snapshotClientOptions(
  options: XenoteerClientOptions,
): Readonly<XenoteerClientOptions> {
  const protocolRange = snapshotProtocolRange(options.protocolRange);
  const clientName = validateDescriptor(
    options.clientName ?? "@xenoteer/sdk",
    "clientName",
  );
  const clientVersion = validateDescriptor(
    options.clientVersion ?? "0.1.0",
    "clientVersion",
  );
  const reconnect = snapshotReconnect(options.reconnect);
  const snapshot = {
    baseUrl: options.baseUrl,
    token: options.token,
    ...(options.requestTimeoutMs === undefined
      ? {}
      : { requestTimeoutMs: options.requestTimeoutMs }),
    ...(options.maxResponseBytes === undefined
      ? {}
      : { maxResponseBytes: options.maxResponseBytes }),
    ...(options.maxArtifactBytes === undefined
      ? {}
      : { maxArtifactBytes: options.maxArtifactBytes }),
    ...(options.fetch === undefined ? {} : { fetch: options.fetch }),
    ...(options.log === undefined ? {} : { log: options.log }),
    clientName,
    clientVersion,
    ...(protocolRange === undefined ? {} : { protocolRange }),
    ...(options.webSocketFactory === undefined
      ? {}
      : { webSocketFactory: options.webSocketFactory }),
    ...(reconnect === undefined ? {} : { reconnect }),
    webSocketQueueCapacity: validateInteger(
      options.webSocketQueueCapacity ?? 256,
      1,
      4096,
      "WebSocket queue capacity",
    ),
    webSocketMaxMessageBytes: validateInteger(
      options.webSocketMaxMessageBytes ?? 1_048_576,
      1024,
      16 * 1_048_576,
      "WebSocket message limit",
    ),
    webSocketHandshakeTimeoutMs: validateInteger(
      options.webSocketHandshakeTimeoutMs ?? 10_000,
      1,
      60_000,
      "WebSocket handshake timeout",
    ),
    webSocketAcknowledgmentTimeoutMs: validateInteger(
      options.webSocketAcknowledgmentTimeoutMs ?? 10_000,
      1,
      60_000,
      "WebSocket acknowledgment timeout",
    ),
    webSocketHeartbeatGraceMs: validateInteger(
      options.webSocketHeartbeatGraceMs ?? 5_000,
      1,
      120_000,
      "WebSocket heartbeat grace",
    ),
    webSocketCloseTimeoutMs: validateInteger(
      options.webSocketCloseTimeoutMs ?? 2_000,
      1,
      30_000,
      "WebSocket close timeout",
    ),
  } satisfies XenoteerClientOptions;
  return Object.freeze(snapshot);
}

/** Safe diagnostic projection. This intentionally cannot reveal the token source. */
export function redactedClientOptions(
  options: XenoteerClientOptions,
): Readonly<Record<string, unknown>> {
  return Object.freeze({
    baseUrl: options.baseUrl,
    token: "<redacted>",
    requestTimeoutMs: options.requestTimeoutMs ?? 35_000,
    maxResponseBytes: options.maxResponseBytes ?? 1_048_576,
    maxArtifactBytes: options.maxArtifactBytes ?? 33_554_432,
    clientName: options.clientName ?? "@xenoteer/sdk",
    clientVersion: options.clientVersion ?? "0.1.0",
    log: options.log === undefined ? undefined : "<configured>",
    webSocketFactory: options.webSocketFactory === undefined
      ? undefined
      : "<configured>",
    reconnect: options.reconnect === undefined
      ? undefined
      : {
          maxAttempts: options.reconnect.maxAttempts ?? 5,
          initialDelayMs: options.reconnect.initialDelayMs ?? 100,
          maxDelayMs: options.reconnect.maxDelayMs ?? 5_000,
        },
    webSocketQueueCapacity: options.webSocketQueueCapacity ?? 256,
    webSocketMaxMessageBytes: options.webSocketMaxMessageBytes ?? 1_048_576,
    webSocketHandshakeTimeoutMs: options.webSocketHandshakeTimeoutMs ?? 10_000,
    webSocketAcknowledgmentTimeoutMs:
      options.webSocketAcknowledgmentTimeoutMs ?? 10_000,
    webSocketHeartbeatGraceMs: options.webSocketHeartbeatGraceMs ?? 5_000,
    webSocketCloseTimeoutMs: options.webSocketCloseTimeoutMs ?? 2_000,
    protocolRange: options.protocolRange ?? {
      major: 1,
      minMinor: 0,
      maxMinor: 0,
    },
  });
}

File: packages/python/src/xenoteer/options.py
# SPDX-License-Identifier: Apache-2.0
"""Connection options with fail-closed credential redaction."""

from __future__ import annotations

import inspect
import ipaddress
import re
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import TypeAlias
from urllib.parse import urlsplit

from .errors import XenoteerError


TokenProvider: TypeAlias = Callable[[], str | Awaitable[str]]
TokenSource: TypeAlias = str | TokenProvider
_TOKEN68 = re.compile(r"[A-Za-z0-9._~+/-]+={0,}\Z")


class BearerToken:
    """Opaque bearer credential whose repr and str are always redacted."""

    __slots__ = ("__value",)

    def __init__(self, value: str) -> None:
        if (
            not isinstance(value, str)
            or not 32 <= len(value) <= 1024
            or _TOKEN68.fullmatch(value) is None
            or re.search(r"=[^=]", value) is not None
        ):
            raise XenoteerError("invalid_token", "invalid Xenoteer bearer token")
        self.__value = value

    def authorization_header(self) -> str:
        """Return the credential only at the HTTP authentication boundary."""

        return f"Bearer {self.__value}"

    def __repr__(self) -> str:
        return "BearerToken(<redacted>)"

    __str__ = __repr__


@dataclass(frozen=True, slots=True)
class ProtocolRange:
    """Inclusive supported minor range within one protocol major."""

    major: int = 1
    min_minor: int = 0
    max_minor: int = 0

    def __post_init__(self) -> None:
        values = (self.major, self.min_minor, self.max_minor)
        if (
            all(isinstance(value, int) and not isinstance(value, bool) for value in values)
            and self.min_minor > self.max_minor
        ):
            raise XenoteerError(
                "reversed_minor_range", "client protocol range is reversed"
            )
        if (
            any(isinstance(value, bool) or not isinstance(value, int) for value in values)
            or any(value < 0 or value > 65_535 for value in values)
        ):
            raise XenoteerError("invalid_request", "client protocol range is invalid")


@dataclass(frozen=True, slots=True)
class ClientOptions:
    """Validated client configuration whose diagnostics cannot reveal a token."""

    base_url: str
    token: TokenSource = field(repr=False)
    request_timeout: float = 35.0
    max_response_bytes: int = 1_048_576
    client_name: str = "xenoteer"
    client_version: str = "0.1.0"
    protocol_range: ProtocolRange = field(default_factory=ProtocolRange)
    heartbeat_interval: float = 15.0
    read_stale_timeout: float | None = None
    max_reconnect_attempts: int = 5

    def __post_init__(self) -> None:
        try:
            parsed = urlsplit(self.base_url)
            parsed.port
        except (TypeError, ValueError):
            raise XenoteerError(
                "invalid_base_url", "base URL must be an absolute HTTP(S) origin"
            ) from None
        if (
            parsed.scheme not in {"http", "https"}
            or not parsed.hostname
            or parsed.username is not None
            or parsed.password is not None
            or parsed.path not in {"", "/"}
            or parsed.query
            or parsed.fragment
        ):
            raise XenoteerError(
                "invalid_base_url",
                "base URL must be an HTTP(S) origin without credentials, path, query, or fragment",
            )
        if parsed.scheme == "http":
            try:
                address = ipaddress.ip_address(parsed.hostname)
            except ValueError:
                raise XenoteerError(
                    "invalid_base_url",
                    "plaintext HTTP is permitted only for a numeric loopback address",
                ) from None
            if not address.is_loopback:
                raise XenoteerError(
                    "invalid_base_url",
                    "plaintext HTTP is permitted only for a numeric loopback address",
                )
        if (
            isinstance(self.request_timeout, bool)
            or not isinstance(self.request_timeout, (int, float))
            or not 0 < self.request_timeout <= 300
        ):
            raise XenoteerError("invalid_request", "request timeout must be in (0, 300] seconds")
        if (
            isinstance(self.max_response_bytes, bool)
            or not isinstance(self.max_response_bytes, int)
            or not 1 <= self.max_response_bytes <= 32 * 1_048_576
        ):
            raise XenoteerError("invalid_request", "response limit is outside its supported range")
        for label, value in (
            ("client name", self.client_name),
            ("client version", self.client_version),
        ):
            if not isinstance(value, str) or not value or len(value.encode()) > 128:
                raise XenoteerError("invalid_request", f"{label} is invalid")
        if (
            isinstance(self.heartbeat_interval, bool)
            or not isinstance(self.heartbeat_interval, (int, float))
            or not 1 <= self.heartbeat_interval <= 300
        ):
            raise XenoteerError(
                "invalid_request", "heartbeat interval must be in [1, 300] seconds"
            )
        if self.read_stale_timeout is not None and (
            isinstance(self.read_stale_timeout, bool)
            or not isinstance(self.read_stale_timeout, (int, float))
            or not self.heartbeat_interval < self.read_stale_timeout <= 600
        ):
            raise XenoteerError(
                "invalid_request",
                "read stale timeout must exceed heartbeat and be at most 600 seconds",
            )
        if (
            isinstance(self.max_reconnect_attempts, bool)
            or not isinstance(self.max_reconnect_attempts, int)
            or not 0 <= self.max_reconnect_attempts <= 20
        ):
            raise XenoteerError(
                "invalid_request", "max reconnect attempts must be in 0..20"
            )

    def safe_dict(self) -> MappingProxyType[str, object]:
        """Return the only supported diagnostics projection."""

        return MappingProxyType(
            {
                "base_url": self.base_url,
                "token": "<redacted>",
                "request_timeout": self.request_timeout,
                "max_response_bytes": self.max_response_bytes,
                "client_name": self.client_name,
                "client_version": self.client_version,
                "protocol_range": self.protocol_range,
                "heartbeat_interval": self.heartbeat_interval,
                "read_stale_timeout": self.read_stale_timeout,
                "max_reconnect_attempts": self.max_reconnect_attempts,
            }
        )

    def __repr__(self) -> str:
        return f"ClientOptions({dict(self.safe_dict())!r})"


async def resolve_token(source: TokenSource) -> BearerToken:
    """Resolve a sync/async provider without retaining provider failures."""

    try:
        value = source() if callable(source) else source
        if inspect.isawaitable(value):
            value = await value
    except Exception:
        raise XenoteerError("invalid_token", "Xenoteer token provider failed") from None
    if not isinstance(value, str):
        raise XenoteerError("invalid_token", "Xenoteer token provider returned a non-string")
    return BearerToken(value)
